synapse/handlers/federation_event.py

```diff
@@ -953,6 +953,47 @@ async def _process_pulled_event(
             )
             return
 
+        # Check if we've already tried to process this event at some point in the past.
+        # We aren't concerned with the expontntial backoff here, just whether it has
+        # failed before.
+        failed_pull_attempt_info = await self._store.get_event_failed_pull_attempt_info(
+            event.room_id, event_id
+        )
+        if failed_pull_attempt_info:
+            # Process previously failed backfill events in the background to not waste
+            # time on something that is bound to fail again.
+            #
+            # TODO: Are we concerned with processing too many events in parallel since
+            # we just fire and forget this off to the background? Should we instead have
+            # a background queue to chew through?
```

Comment:
Should we be concerned?

I guess someone could spam push us with the same failing event over and over and we could queue them all up in parallel exhausting our resources. They would run into the federation rate limiters though.

And we won't backfill the same event over and over thanks to https://github.com/matrix-org/synapse/pull/13635

Reply:
We already [`concurrently_execute(process_pdus_for_room, ...)` when servers push us events](https://github.com/matrix-org/synapse/blob/c97198ee140fabf2c7444c9beea5baa87e98fd99/synapse/federation/federation_server.py#L522-L524) so is this really that different? (it does have a ceiling though, `TRANSACTION_CONCURRENCY_LIMIT = 10` )

One way we can side-step having to think about this is only applying it to backfill where we don't have to worry.

Reply:
We could handle things a layer up in `process_pulled_events()` where we could `run_as_background_process(...)` with the for-loop that iterates over the events.

Reply:
Updated in https://github.com/matrix-org/synapse/pull/15585/commits/45934fe0fb218e6cfbfa5dda3cef6be7930b8e13 :fast_forward: